### PR TITLE
Add a customization point for put_parcel so we can override actions (…

### PR DESCRIPTION
**Pull request by biddisco**
…e.g. rdma)

### CODE REVIEW

File: hpx/traits/action_put_parcel.hpp [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+{
+    namespace detail {
+
+        ///////////////////////////////////////////////////////////////////////
+        template <typename Action>
+        inline naming::address&& complement_addr(naming::address& addr)
+        {
+            if (components::component_invalid == addr.type_)
+            {
+                addr.type_ = components::get_component_type<
```
Comment (sithhell) [inline hpx/traits/action_put_parcel.hpp]: The setting of component type should be done in the put_parcel functions already.

**Comment by biddisco** [comment]
The specialization at this level was intended to bypas the generation of a 'parcel' altogether. I wanted to intercept things before put_parcel, but since I've not actually implemented the rdma actions yet, I'm open to these suggestions. I will play some more and start implementing it and see where it leads...

**Comment by sithhell** [comment]
While thinking about this a little more, we already have this specialization in place: It's the message_handler that can be overriden for a given kind of action (we already do this for parcel coalescing). This message handler could then perform the necessary rdma action given the specific parcel (the creation of a parcel mainly involves the construction of a transfer_action).

**Comment by hkaiser** [comment]
> While thinking about this a little more, we already have this specialization in place: It's the message_handler that can be overriden for a given kind of action (we already do this for parcel coalescing). This message handler could then perform the necessary rdma action given the specific parcel (the creation of a parcel mainly involves the construction of a transfer_action).

The advantage of the proposed customization point is that it avoids to create a parcel at all.  Also, we already know that the message handler hook is very inefficient as it has to look up the action based on its name (string). why add this overhead to the RDMA operations if not needed. This kind of defeats the purpose.

**Comment by sithhell** [comment]
Well, improving the message handler hook might be something to look into then. I am just saying that this might be the perfect place to hook those RDMA operations in. It is about handling messages in a special way. The overhead of creating a parcel should be very minimal (just one allocation for the transfer action), despite the advantages you have when having a parcel: calculated message size, zero copy buffers and the target GID properly reference counted.

**Comment by hkaiser** [comment]
> Well, improving the message handler hook might be something to look into then

Yes, this is on my list of things to do. Doing the hooking based on a type trait instead of the action name looks like the right thing to so.

However, handling the RDMA actions further down the chain requires to create a parcel instance which is not necessary in the first place. We don't need any transfer actions either as we want to circumvent the need for any serialization. That means there is no need to calculate the message sizes (would be trivial for the very selected set of arguments we would support anyways), no need to have zero-copy buffers either. I admit, I don't completely understand the consequences wrt the target GID, though.

**Comment by hkaiser** [comment]
@sithhell would you be ok now to go ahead and merge this?

File: hpx/traits/action_put_parcel.hpp (hunk 1 above)
Comment (sithhell) [inline hpx/traits/action_put_parcel.hpp]: No, that means it should happen one layer above.

**Comment by hkaiser** [comment]
@sithhell, @biddisco Should this still go into the V1.0 release?

**Comment by msimberg** [comment]
@sithhell, @biddisco Repeating @hkaiser's question, but for v1.1.0.

**Comment by biddisco** [comment]
I use this customization point in my testing with RMA stuff, so I'd like to see it added to 1.2.0

**Comment by hkaiser** [comment]
@biddisco, @sithhell what should happen to this PR? Will you be able to agree on a path forward? Otherwise I'd suggest to abandon the proposed changes.

**Comment by biddisco** [comment]
As mentioned before, I have been using this customization point in my RDMA branch, so when that get's merged, I'll revist the problem. For now, we can delete this PR and forget about it.

**Pull request closed without merging**